### PR TITLE
Revising fluent launch arguments

### PR DESCRIPTION
**Pull request by raph-luc**
Fixes https://github.com/ansys/pyfluent/issues/1769

Removing support for `launch_fluent(server_info_filepath=...)` as it does not make much sense to support it considering that the file will be deleted by Fluent when it exits, regardless of what we do in PyFluent.

Raising `ValueError` exceptions if arguments are provided in unsupported combinations. Might've missed a few but I think I at least got the most important ones, please let me know (or directly commit) if you notice any missed combinations.

### CODE REVIEW

File: src/ansys/fluent/core/launcher/launcher.py [hunk 1]
```diff
@@ -615,20 +611,69 @@ def launch_fluent(
...
+            "topy",
+            "case_filepath",
+            "lightweight_mode",
+            "journal_filepath",
+            "case_data_filepath",
+        ]
+        invalid_arg_names = []
+        for arg_name in arg_names:
+            if argvals[arg_name] is not None:
+                invalid_arg_names.append(arg_name)
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/launcher/launcher.py:659]:
```suggestion
        invalid_arg_names=filter(lambda arg_name:argvals[arg_name] is not None, arg_names)
```

(for example)

Comment (raph-luc) [reply src/ansys/fluent/core/launcher/launcher.py]: I knew there had to be a better way to do that. Wasn't familiar with `filter()`, that is going to be useful, thanks Sean

Comment (mkundu1) [reply src/ansys/fluent/core/launcher/launcher.py]: `invalid_arg_names = [arg_name for arg_name in arg_names if argvals[arg_name] is not None]`

if you prefer list-comprehension

Comment (raph-luc) [reply src/ansys/fluent/core/launcher/launcher.py]: Using list comprehension is also better than what I suggested initially, thanks Mainak. 

Between filter and list comprehension, either one works fine I think, going to stick with filter for this one as it is an interesting built-in function :)

File: src/ansys/fluent/core/launcher/launcher.py [hunk 2]
```diff
@@ -615,20 +611,68 @@ def launch_fluent(
...
+            "'start_watchdog' argument for 'launch_fluent' is currently not supported "
+            "when starting a remote Fluent PyPIM client."
+        )
+
+    if (
+        start_watchdog is None
+        and cleanup_on_exit
+        and (
+            fluent_launch_mode == LaunchMode.CONTAINER
+            or fluent_launch_mode == LaunchMode.STANDALONE
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/launcher/launcher.py:633]:
```suggestion
        and (fluent_launch_mode in (LaunchMode.CONTAINER, LaunchMode.STANDALONE)
```

File: src/ansys/fluent/core/launcher/launcher.py [hunk 3]
```diff
@@ -615,20 +611,68 @@ def launch_fluent(
 
     del start_container
 
-    if start_watchdog is None and cleanup_on_exit:
+    if additional_arguments is None:
+        additional_arguments = ""
+    elif fluent_launch_mode == LaunchMode.PIM:
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/launcher/launcher.py:616]: In later changes, fluent_launch_mode could be a key `launcher_factory.create(key)` then each launcher could be responsible for constraining/validating its own launch configuration in `launcher.launch(<args>)`. E.g., `pim_launcher.launch(<args>)` would not have an `additional_arguments` argument.

File: src/ansys/fluent/core/launcher/launcher.py [hunk 4]
```diff
@@ -615,20 +611,68 @@ def launch_fluent(
...
+        raise ValueError(
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/launcher/launcher.py:617]: these are normally `TypeError`. 
```
>>> def foo(): pass
...
>>>
>>> foo(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes 0 positional arguments but 1 was given
>>> foo(x=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() got an unexpected keyword argument 'x'
>>>
```

File: src/ansys/fluent/core/launcher/launcher.py (hunk 1 above)
Comment (seanpearsonuk) [reply src/ansys/fluent/core/launcher/launcher.py]: Bear in mind that there is a performance difference. One of these is almost twice as quick a the other:
```
>>> xs = range(100000000)
>>> ys = list(filter(lambda x: x%2, xs))
>>> zs = [x for x in xs if x%2]
```
but it doesn't really matter here.

File: src/ansys/fluent/core/launcher/launcher.py (hunk 4 above)
Comment (raph-luc) [reply src/ansys/fluent/core/launcher/launcher.py:617]: In this case the "type" (`NoneType` or `str`) is correct as the function does accept and expect that variable with those types. However, with this combination of arguments, the "value" is incorrect as it needs to be `None`. I think `ValueError` makes a bit more sense than `TypeError`.

Googling this, most people seem to suggest `ValueError` as well, first Google links that show up here:
https://stackoverflow.com/questions/256222/which-exception-should-i-raise-on-bad-illegal-argument-combinations-in-python

https://www.adamsmith.haus/python/answers/how-to-raise-an-exception-for-illegal-argument-combinations-in-python

Comment (seanpearsonuk) [reply src/ansys/fluent/core/launcher/launcher.py:617]: Okay, that makes sense, yes.

File: src/ansys/fluent/core/launcher/launcher.py [hunk 5]
```diff
@@ -615,20 +611,68 @@ def launch_fluent(
...
-        if not fluent_launch_mode == LaunchMode.CONTAINER:
+    if dry_run and fluent_launch_mode != LaunchMode.CONTAINER:
+        raise ValueError(
+            "'dry_run' argument for 'launch_fluent' currently is only "
+            "supported when starting containers."
+        )
+
+    argvals = locals()
+
+    if fluent_launch_mode != LaunchMode.STANDALONE:
```
Comment (seanpearsonuk) [inline src/ansys/fluent/core/launcher/launcher.py:646]: We can really see how the launcher factory is going to smash this! I will create a detailed gh issue.

**Comment by raph-luc** [comment]
@seanpearsonuk should I close or wait on this PR due to the incoming launch fluent rework which is probably going to make most of this obsolete? Or should we have that rework perhaps based on this branch?

**Comment by seanpearsonuk** [comment]
> @seanpearsonuk should I close or wait on this PR due to the incoming launch fluent rework which is probably going to make most of this obsolete? Or should we have that rework perhaps based on this branch?

I would be very happy for you to merge this, assuming there are no outstanding points.

**Review by hpohekar — APPROVED** [review]

**Review by mkundu1 — APPROVED** [review]

**Review by seanpearsonuk — APPROVED** [review]

**Pull request merged**